services/spar/src/Spar/Data.hs

```diff
@@ -0,0 +1,249 @@
+import Control.Monad.Identity
+import Data.Int
+import Data.List.NonEmpty as NE
+import Data.Maybe (catMaybes)
+import Data.Misc ((<$$>))
+import Data.String.Conversions
+import Data.Time
+import Data.X509 (SignedCertificate)
+import GHC.Stack
+import Lens.Micro
+import Spar.Data.Instances ()
+import Spar.Options as Options
+import Spar.Types
+import System.Logger (Logger)
+import URI.ByteString
+import Util.Options (casEndpoint, casKeyspace, epHost, epPort)
+
+import qualified Cassandra.Schema as Cas
+import qualified Cassandra.Settings as Cas
+import qualified Data.Id as Brig
+import qualified Data.UUID as UUID
+import qualified SAML2.WebSSO as SAML
+import qualified Spar.Options as Opts
+import qualified System.Logger as Log
+
+
+----------------------------------------------------------------------
+-- init
+
+schemaVersion :: Int32
+schemaVersion = 0
+
+initCassandra :: Opts.Opts -> Logger -> IO ClientState
+initCassandra opts lgr = do
+    connectString <- maybe (return $ NE.fromList [cs $ Opts.cassandra opts ^.casEndpoint.epHost])
+               (Cas.initialContacts "cassandra_spar")
+               (cs <$> Opts.discoUrl opts)
+    cas <- Cas.init (Log.clone (Just "cassandra.spar") lgr) $ Cas.defSettings
+      & Cas.setContacts (NE.head connectString) (NE.tail connectString)
+      & Cas.setPortNumber (fromIntegral $ Options.cassandra opts ^. casEndpoint . epPort)
+      & Cas.setKeyspace (Keyspace $ Options.cassandra opts ^. casKeyspace)
+      & Cas.setMaxConnections 4
+      & Cas.setMaxStreams 128
+      & Cas.setPoolStripes 4
+      & Cas.setSendTimeout 3
+      & Cas.setResponseTimeout 10
+      & Cas.setProtocolVersion V3
+    runClient cas $ Cas.versionCheck schemaVersion
+    pure cas
+
+
+----------------------------------------------------------------------
+-- helpers
+
+data Env = Env
+  { dataEnvNow :: UTCTime
+  , dataEnvMaxTTLAuthRequests :: TTL "authreq"
+  , dataEnvMaxTTLAssertions   :: TTL "authresp"
+  }
+  deriving (Eq, Show)
+
+data TTLError = TTLTooLong | TTLInPast
+  deriving (Eq, Show)
+
+mkTTLAuthnRequests :: MonadError TTLError m => Env -> UTCTime -> m (TTL "authreq")
+mkTTLAuthnRequests (Env now maxttl _) = mkTTL now maxttl
+
+mkTTLAssertions :: MonadError TTLError m => Env -> UTCTime -> m (TTL "authresp")
+mkTTLAssertions (Env now _ maxttl) = mkTTL now maxttl
+
+mkTTL :: MonadError TTLError m => UTCTime -> TTL a -> UTCTime -> m (TTL a)
+mkTTL now maxttl endOfLife = if
+  | actualttl > maxttl -> throwError TTLTooLong
+  | actualttl <= 0     -> throwError TTLInPast
+  | otherwise          -> pure actualttl
+  where
+    actualttl = TTL . round @Double . realToFrac $ endOfLife `diffUTCTime` now
+
+err2err :: (m ~ Either TTLError, MonadThrow m') => m a -> m' a
+err2err = either (throwM . ErrorCall . show) pure
+
+
+----------------------------------------------------------------------
+-- saml state handling
+
+storeRequest :: (HasCallStack, MonadClient m) => Env -> SAML.ID SAML.AuthnRequest -> SAML.Time -> m ()
```

Comment:
Would it not look clear if instead of passing `Env` around we would have a `HasEnv m` constraint or similar?